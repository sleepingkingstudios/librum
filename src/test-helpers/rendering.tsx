import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import {
  MemoryRouter,
  Routes,
  Route
} from "react-router-dom";
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import {
  createStore,
  storeType,
} from './store';
import {
  ThemeContext,
  defaultTheme,
} from '@themes/context';
import type { Theme } from '@themes';

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true | withRouterOptions;
  store?: (storeType | true);
  theme?: (Theme | true);
}

interface withRouterOptions {
  initialEntries?: Array<string>;
  path?: string;
}

export const withRouter = (
  component: React.ReactElement,
  {
    initialEntries = ['/'],
    path,
  }: withRouterOptions,
): React.ReactElement => {
  let wrapped = component;

  if (path) {
    wrapped = (
      <Routes>
        <Route path={path} element={wrapped} />
      </Routes>
    );
  }

  return (
    <MemoryRouter initialEntries={initialEntries}>
      { wrapped }
    </MemoryRouter>
  );
};

export const withStore = (
  component: React.ReactElement,
  { store }: { store: (storeType | true) }
): React.ReactElement => {
  let configuredStore: storeType;

  if (store === true) {
    configuredStore = createStore().store;
  } else {
    configuredStore = store;
  }

  return (
    <ReduxProvider store={configuredStore}>
      { component }
    </ReduxProvider>
  );
};

export const withTheme = (
  component: React.ReactElement,
  { theme }: { theme: (Theme | true) }
): React.ReactElement => {
  let configuredTheme: Theme;

  if (theme === true) {
    configuredTheme = defaultTheme;
  } else {
    configuredTheme = theme;
  }

  return (
    <ThemeContext.Provider value={configuredTheme}>
      { component }
    </ThemeContext.Provider>
  );
};

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions,
  strictMode = true,
): RenderResult => {
  let wrapped: React.ReactElement = component;

  if (strictMode) {
    wrapped = (<React.StrictMode>{ wrapped }</React.StrictMode>);
  }

  if (!options) { return render(wrapped); }

  if ('router' in options && options.router) {
    if (options.router === true) {
      wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
    } else {
      wrapped = withRouter(
        wrapped,
        {
          initialEntries: options.initialEntries,
          ...options.router,
        },
      );
    }
  }

  if ('store' in options) {
    wrapped = withStore(wrapped, { store: options.store });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
