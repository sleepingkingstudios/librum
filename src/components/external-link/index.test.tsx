import * as React from 'react';

import '@testing-library/jest-dom';
import { render } from '@test-helpers/rendering';

import { ExternalLink } from './index';

describe('<ExternalLink />', () => {
  describe('with children: value', () => {
    const path = '/path/to/resource';
    const children = (<strong>Link Text</strong>);

    it('should generate the link', () => {
      const { getByRole } = render(
        <ExternalLink to={path}>
          { children }
        </ExternalLink>,
        { router: true },
      );

      const link = getByRole('link');
      expect(link).toBeVisible();
      expect(link).toHaveAttribute('href', path);
      expect(link).toHaveTextContent('Link Text');
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <ExternalLink to={path}>
          { children }
        </ExternalLink>,
        { router: true },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with className: value', () => {
    const path = '/path/to/resource';

    it('should generate the link', () => {
      const { getByRole } = render(
        <ExternalLink className="custom-link" to={path} />,
        { router: true },
      );

      const link = getByRole('link');
      expect(link).toBeVisible();
      expect(link).toHaveAttribute('href', path);
      expect(link).toHaveTextContent(path);
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <ExternalLink className="custom-link" to={path} />,
        { router: true },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with label: value', () => {
    const path = '/path/to/resource';

    it('should generate the link', () => {
      const { getByRole } = render(
        <ExternalLink label="Link Text" to={path} />,
        { router: true },
      );

      const link = getByRole('link');
      expect(link).toBeVisible();
      expect(link).toHaveAttribute('href', path);
      expect(link).toHaveTextContent('Link Text');
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <ExternalLink label="Link Text" to={path} />,
        { router: true },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with to: null', () => {
    it('should return null', () => {
      const { container } = render(
        <ExternalLink to={null} />,
        { router: true },
      );

      expect(container.firstChild).toBeNull();
    });

    describe('with children: value', () => {
      const children = (<strong>Link Text</strong>);

      it('should render the contents', () => {
        const { getByText } = render(
          <ExternalLink to={null}>
            { children }
          </ExternalLink>,
          { router: true },
        );

        expect(getByText('Link Text')).toBeVisible();
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <ExternalLink to={null}>
            { children }
          </ExternalLink>,
          { router: true },
        );

        expect(asFragment).toMatchSnapshot();
      });
    });

    describe('with className: value', () => {
      it('should return null', () => {
        const { container } = render(
          <ExternalLink className="custom-link" to={null} />,
          { router: true },
        );

        expect(container.firstChild).toBeNull();
      });
    });

    describe('with label: value', () => {
      it('should render the contents', () => {
        const { getByText } = render(
          <ExternalLink to={null} label="Link Text" />,
          { router: true },
        );

        expect(getByText('Link Text')).toBeVisible();
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <ExternalLink to={null} label="Link Text" />,
          { router: true },
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe('with to: an empty string', () => {
    it('should return null', () => {
      const { container } = render(
        <ExternalLink to={''} />,
        { router: true },
      );

      expect(container.firstChild).toBeNull();
    });

    describe('with children: value', () => {
      const children = (<strong>Link Text</strong>);

      it('should render the contents', () => {
        const { getByText } = render(
          <ExternalLink to={''}>
            { children }
          </ExternalLink>,
          { router: true },
        );

        expect(getByText('Link Text')).toBeVisible();
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <ExternalLink to={''}>
            { children }
          </ExternalLink>,
          { router: true },
        );

        expect(asFragment).toMatchSnapshot();
      });
    });

    describe('with className: value', () => {
      it('should return null', () => {
        const { container } = render(
          <ExternalLink className="custom-link" to={''} />,
          { router: true },
        );

        expect(container.firstChild).toBeNull();
      });
    });

    describe('with label: value', () => {
      it('should render the contents', () => {
        const { getByText } = render(
          <ExternalLink to={''} label="Link Text" />,
          { router: true },
        );

        expect(getByText('Link Text')).toBeVisible();
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <ExternalLink to={''} label="Link Text" />,
          { router: true },
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe('with to: an absolute path', () => {
    const path = '/path/to/resource';

    it('should generate the link', () => {
      const { getByRole } = render(
        <ExternalLink to={path} />,
        { router: true },
      );

      const link = getByRole('link');
      expect(link).toBeVisible();
      expect(link).toHaveAttribute('href', path);
      expect(link).toHaveTextContent(path);
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <ExternalLink to={path} />,
        { router: true },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with to: a relative path', () => {
    const path = 'path/to/resource';

    it('should generate the link', () => {
      const { getByRole } = render(
        <ExternalLink to={path} />,
        { router: true },
      );

      const link = getByRole('link');
      expect(link).toBeVisible();
      expect(link).toHaveAttribute('href', `/${path}`);
      expect(link).toHaveTextContent(path);
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <ExternalLink to={path} />,
        { router: true },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with to: a url', () => {
    const path = 'example.com/path/to/resource';

    it('should generate the link', () => {
      const { getByRole } = render(
        <ExternalLink to={path} />,
        { router: true },
      );

      const link = getByRole('link');
      expect(link).toBeVisible();
      expect(link).toHaveAttribute('href', `https://${path}`);
      expect(link).toHaveTextContent(path);
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <ExternalLink to={path} />,
        { router: true },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with to: a url with a protocol', () => {
    const path = 'http://example.com/path/to/resource';

    it('should generate the link', () => {
      const { getByRole } = render(
        <ExternalLink to={path} />,
        { router: true },
      );

      const link = getByRole('link');
      expect(link).toBeVisible();
      expect(link).toHaveAttribute('href', path);
      expect(link).toHaveTextContent(path);
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <ExternalLink to={path} />,
        { router: true },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with to: a url with a subdomain', () => {
    const path = 'www.example.com/path/to/resource';

    it('should generate the link', () => {
      const { getByRole } = render(
        <ExternalLink to={path} />,
        { router: true },
      );

      const link = getByRole('link');
      expect(link).toBeVisible();
      expect(link).toHaveAttribute('href', `https://${path}`);
      expect(link).toHaveTextContent(path);
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <ExternalLink to={path} />,
        { router: true },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
