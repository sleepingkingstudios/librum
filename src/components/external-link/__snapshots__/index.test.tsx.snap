// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExternalLink /> with children: value should match the snapshot 1`] = `
<DocumentFragment>
  <a
    href="/path/to/resource"
  >
    <span>
      <strong>
        Link Text
      </strong>
    </span>
  </a>
</DocumentFragment>
`;

exports[`<ExternalLink /> with className: value should match the snapshot 1`] = `
<DocumentFragment>
  <a
    href="/path/to/resource"
  >
    <span
      class="custom-link"
    >
      /path/to/resource
    </span>
  </a>
</DocumentFragment>
`;

exports[`<ExternalLink /> with label: value should match the snapshot 1`] = `
<DocumentFragment>
  <a
    href="/path/to/resource"
  >
    <span>
      Link Text
    </span>
  </a>
</DocumentFragment>
`;

exports[`<ExternalLink /> with to: a relative path should match the snapshot 1`] = `
<DocumentFragment>
  <a
    href="/path/to/resource"
  >
    <span>
      path/to/resource
    </span>
  </a>
</DocumentFragment>
`;

exports[`<ExternalLink /> with to: a url should match the snapshot 1`] = `
<DocumentFragment>
  <a
    href="https://example.com/path/to/resource"
  >
    <span>
      example.com/path/to/resource
    </span>
  </a>
</DocumentFragment>
`;

exports[`<ExternalLink /> with to: a url with a protocol should match the snapshot 1`] = `
<DocumentFragment>
  <a
    href="http://example.com/path/to/resource"
  >
    <span>
      http://example.com/path/to/resource
    </span>
  </a>
</DocumentFragment>
`;

exports[`<ExternalLink /> with to: a url with a subdomain should match the snapshot 1`] = `
<DocumentFragment>
  <a
    href="https://www.example.com/path/to/resource"
  >
    <span>
      www.example.com/path/to/resource
    </span>
  </a>
</DocumentFragment>
`;

exports[`<ExternalLink /> with to: an absolute path should match the snapshot 1`] = `
<DocumentFragment>
  <a
    href="/path/to/resource"
  >
    <span>
      /path/to/resource
    </span>
  </a>
</DocumentFragment>
`;

exports[`<ExternalLink /> with to: an empty string with children: value should match the snapshot 1`] = `[Function]`;

exports[`<ExternalLink /> with to: an empty string with label: value should match the snapshot 1`] = `
<DocumentFragment>
  <span>
    Link Text
  </span>
</DocumentFragment>
`;

exports[`<ExternalLink /> with to: null with children: value should match the snapshot 1`] = `[Function]`;

exports[`<ExternalLink /> with to: null with label: value should match the snapshot 1`] = `
<DocumentFragment>
  <span>
    Link Text
  </span>
</DocumentFragment>
`;
